src/lib/topology.c

```diff
@@ -0,0 +1,1049 @@
+/*
```

Comment:
@singalsu same for the parser too. The major difference between the two is one takes the the topology file as input and the other uses the bytes.

Reply:
@ranj063 Yes, I'd prefer this way to keep the data in memory for both cases since it's much simpler. The fseek() pointer arithmetic is really tricky and needs lot of local copied parts.

Reply:
@singalsu would it be simpler for your parser (and it would be smaller) if the parser was only parsing a file of IPC structures instead of the ALSA structure & tokens ? i.e.

 1) take the ALSA binary file strip out ALSA struct and convert to IPC structs
 2) load the new file with IPC structs into SOF.
 
Step 1 above is mostly done already in your patch and in the driver (topology.c) and means creating a small cmd line tool based on topology.c that does this conversion in userspace. This way your FW parser is far simpler as it will just call handler.c calls directly and pass in each structure ?